### DISCUSSION TITLE
PWM for Analog strips

### DISCUSSION THREAD

**Discussion by marcobrianza** [forkineye/ESPixelStick#735] [Q&A]
Hi have seen that PWM output is possible with a special I2C chip.
Is it also possible using the processor pins?

**Comment by MartinMueller2003** [comment]
Yes. If you compile with the relay functionality turned in, you can output PWM on a few of the ESPs output pins. Not all platforms support the outputs so you may need to do a custom build.

**Reply by marcobrianza** [reply]
Thank you Martin, I have seen that on LOIN 32 PRO and also on the wemos D1 mini, the feature is already there in one of the latest CI release (no build required). AWESOME!  
On ESP32 the GPIO ID is set to -1 if not configured.
On ESP8266 it starts with 4294967295 (with error indication) see attachment
![gpio](https://github.com/forkineye/ESPixelStick/assets/1008278/be777bf6-ada8-4dc1-b0a8-d3c78521d042)
I guess it is just a type definition error.

Calling the feature Relay is misleading at this point I would call it GPIO. 
I immagine that  "Trigger Threshold" applies only if PWM is disabled, right?

have a nice weekend.
Marco

**Comment by MartinMueller2003** [comment]
It is -1 on both builds. When it is -1, the channel is not configured. You need to assign a GPIO. Yes the outputs support both PWM and on/off mode. The trigger point is only valid in on/off mode.

**Comment by marcobrianza** [comment]
Did you see the attached file?In esp8266 -1 becomes 4294967295On 27 Jan 2024, at 16:28, Martin Mueller ***@***.***> wrote:﻿
It is -1 on both builds. When it is -1, the channel is not configured. You need to assign a GPIO. Yes the outputs support both PWM and on/off mode. The trigger point is only valid in on/off mode.

**Comment by MartinMueller2003** [comment]
Anything more on this or can we close the discussion?

**Reply by RedLightning42** [reply]
I just received a ESPixelStick v3 and loaded your latest beta firmware through the EXE. I was able to get DMX output working (driving from Effects, ultimate goal is sACN, which I'm testing soon).

But ideally, I want to have the options to do PWM or even Relay (I have extra Relay Boards if needed) to control stuff that already has battery packs or just low voltage LEDs, and I seem to be missing part of the setup to get the options in the image above. Is there a guide yet?

**Comment by MartinMueller2003** [comment]
The ESPixelStick V3 hardware does not support relay functionality so you will not see it in the config options. There are other platforms that do support PWM outputs, just not the V3 hardware.

**Comment by RedLightning42** [comment]
So what you're saying is, I have the wrong hardware for what I'm trying to
achieve. Oh well, guess I should've done more research first...

I did also pick up some ESP Relay Shields, I don't suppose you know if I
can get the Wemos D1 Mini from the top and reconfigure it to ACT as a Relay
controller over e1.31 (sACN)?

**Comment by MartinMueller2003** [comment]
ESPixelStick V4 SW supports almost any ESP platform you can put together. Yes it can do that.

**Comment by RedLightning42** [comment]
Martin,

I hate to bombard you as Tech Support for your firmware, but want to see if
you can spot the solution.

I bought some cheap ESP32 Relay Shields off Amazon, plugged one into the
Wemos D1 from a PixelStick setup, loaded the v4.5 beta firmware for a D1.
Got the Relay on the second output, can even control them from sACN (E1.31)
as shown through the web interface, but the relay shield itself won't
trigger.
I'm thinking it's something in the GPIO ID in the firmware, or maybe the
relay is expecting something different from what it's getting sent, but I'm
really new to ESP32 and all of this, so if you know where to look, I'd
appreciate it.

Thanks!

**Comment by MartinMueller2003** [comment]
Please send a link to the relay board. I suspect you have a voltage problem. ESP GPIOs run at 3 volts and do NOT have a lot of current carrying capacity.

**Comment by RedLightning42** [comment]
https://www.amazon.com/HiLetgo-Relay-Shield-Module-WeMos/dp/B01NACU547/147-6117991-8813133

It says 5v, and I'm powering the Wemos off the 5v USB for now. I'm handy
enough with a soldering iron if I need to reroute a few pins though. Was
already planning on using some extension cables to test a different board
as I'm out of Female Headers at the moment.

**Comment by MartinMueller2003** [comment]
the esp gpio cannot drive that relay directly. You need to add a 3.3c to 5v level shifter.

**Comment by RedLightning42** [comment]
Actually, I was just about to message you. Turns out, if you don't use the
headers, but connect 5v, GND and D4 pins to 5v, GND and D1 (on the Relay)
it works to at least trigger the relay to flip. Part of the problem (found
in an Amazon review) is the relay is expecting D1, but the newer boards use
that for something else. Now I just need to figure out the config (GPIO ID)
to trip D4 on the Wemos, and I bet it works.

**Reply by RedLightning42** [reply]
Just coming back to update the final result. It DOES work! If anyone runs into a similar situation, here's the solution:

Connect the 5v and GRD pins from the WEMOS to the Relay Card.
Connect the WEMOS pin D4 to the Relay Card D1.
Using the ESPixelStick V4 or newer firmware, set your Relay output to GPIO 2 (Invert and test as needed). 

For some reason, the Relay wants to trigger on the D1 pin, but I missed the GPIO pins having different functions and they're not in numerical order. There's probably other combinations that work, but this is what I got functioning. With a custom wiring harness, I now have a single Relay working over sACN from a Theater Light Board. Thanks!

**Comment by MartinMueller2003** [comment]
Even if it works, it is not using PWM AND the trigger voltage is marginal. You may find times when it seems to skip a transaction.
